### DISCUSSION TITLE
Restrictions for Napi::String references

### DISCUSSION THREAD

**Discussion by lboquillon** [nodejs/help#4752] [Q&A]
Hi there,

I hope you're doing well. I've come across an issue while attempting to create a weak reference for an Napi::String as shown in the code snippet below:

```
Napi::String str = ...;
...
Napi::Reference<Napi::String> ref = Napi::Weak(str);
```

Unfortunately, this resulted in an napi_invalid_arg exception. After investigating, it appears that the problem originates from the following section of the source code:

```
napi_status NAPI_CDECL napi_create_reference(napi_env env,
                                             napi_value value,
                                             uint32_t initial_refcount,
                                             napi_ref* result) {
...
  v8::Local<v8::Value> v8_value = v8impl::V8LocalValueFromJsValue(value);
  if (!(v8_value->IsObject() || v8_value->IsFunction() ||
        v8_value->IsSymbol())) {
    return napi_set_last_error(env, napi_invalid_arg);
  }
...
}
```

I was wondering if you could kindly shed some light on the reason behind the restriction that prevents using a Napi::String to create a Napi::Reference. Alternatively, would it be feasible to consider loosing this restriction and allowing strings to be accepted as well? If this change aligns with the project's goals, I'd be more than willing to work on the Pull Request to implement it.

Am I perhaps overlooking something, and there is an alternative method for creating a reference to a string that I'm not aware of?

Thank you for your time and assistance. Looking forward to your guidance on this matter.

Best regards,
Leonardo Boquillon

**Comment by Whitecx** [comment]
While it's not very clear from the node-addon-api docs, here's a quote from a [section](https://github.com/nodejs/node/blob/8f0f17e1e3b6c4e58ce748e06343c5304062c491/doc/api/n-api.md#references-to-values-with-a-lifespan-longer-than-that-of-the-native-method) of the Node-API docs

"_Node-API provides methods for creating persistent references to values. Currently Node-API only allows references to be created for a limited set of value types, including object, external, function, and symbol._"

As far as why this isn't yet supported, I'm not sure 🤷🏿‍♂️. It might be fun to look into though! I'd be open to following up on an issue w/ you if you're interested.

**Reply by Whitecx** [reply]
Here's a thread where some discussion/progress was made on implementing a version of references that would work w/ all types last year -> https://github.com/orgs/nodejs/discussions/41842

**Comment by mhdawson** [comment]
@lboquillon 

> I was wondering if you could kindly shed some light on the reason behind the restriction that prevents using a Napi::String to create a Napi::Reference.

References started as a way to be able to keep alive Objects (and types that are subclasses of Objects) that can be collected by the Garbage collector.  Strings are a primitive type and therefore were not included in what could be supported, possibly because References depend on actions/callbacks available for Objects, I've have to look at the code to be more specific..

As mentioned there has been some discussion about creating some APIs with difference semantics, but it has not been completed yet.
